File: utf64.cs
/*!
 * UTF-64 — Universal Transformation Format 64
 * Specification v0.2 — C# Reference Implementation
 *
 * Changes from v0.1:
 *   - Magic bytes removed from encoder output.
 *     Decoder still accepts (and strips) legacy magic-prefixed streams.
 */

using System;
using System.Collections.Generic;
using System.Text;

public static class Utf64
{
    // ─── Constants ────────────────────────────────────────────────────────────

    private static readonly byte[] LegacyMagic = Encoding.ASCII.GetBytes("&-9EX");

    private static readonly char[] Tier0Chars =
        { ' ', 'e', 't', 'a', 'o', 'i', 'n', 's', 'h', 'r', 'l', '\n' };

    private static readonly Dictionary<char, int> Tier0Map = new();

    private const byte EscCap    = 0xC;
    private const byte EscTier1  = 0xD;
    private const byte EscTier23 = 0xE;
    private const byte EscTier4  = 0xF;
    private const byte CtrlCap   = 0x0;

    static Utf64()
    {
        for (int i = 0; i < Tier0Chars.Length; i++)
            Tier0Map[Tier0Chars[i]] = i;
    }

    // ─── Encoder ──────────────────────────────────────────────────────────────

    private static void EncodeTier3(uint cp, List<byte> nib)
    {
        nib.Add(EscTier23);
        nib.Add(0x8);
        for (int shift = 24; shift >= 0; shift -= 4)
            nib.Add((byte)((cp >> shift) & 0xF));
    }

    private static void EncodeCP(uint cp, List<byte> nib)
    {
        string chars  = char.ConvertFromUtf32((int)cp);
        string lower  = chars.ToLowerInvariant();
        bool isUpper  = chars != lower;
        char lowerCh  = lower[0];
        int t0l       = Tier0Map.TryGetValue(lowerCh, out int v0) ? v0 : -1;

        if (isUpper && t0l >= 0)
        {
            nib.Add(EscCap); nib.Add(CtrlCap); nib.Add((byte)t0l);
            return;
        }

        char ch = chars[0];
        if (Tier0Map.TryGetValue(ch, out int t0))
        {
            nib.Add((byte)t0);
            return;
        }

        if (cp <= 0xFF)
        {
            nib.Add(EscTier1);
            nib.Add((byte)((cp >> 4) & 0xF));
            nib.Add((byte)(cp & 0xF));
            return;
        }

        if (cp <= 0xFFFF)
        {
            byte n3 = (byte)((cp >> 12) & 0xF);
            if (n3 <= 0x7)
            {
                nib.Add(EscTier23);
                nib.Add(n3);
                nib.Add((byte)((cp >> 8) & 0xF));
                nib.Add((byte)((cp >> 4) & 0xF));
                nib.Add((byte)(cp & 0xF));
            }
            else
            {
                EncodeTier3(cp, nib);
            }
            return;
        }

        if (cp <= 0xFFFFFFFF)
        {
            EncodeTier3(cp, nib);
            return;
        }

        // Tier 4
        nib.Add(EscTier4);
        for (int shift = 56; shift >= 0; shift -= 4)
            nib.Add((byte)((cp >> shift) & 0xF));
    }

    public static byte[] Encode(string s)
    {
        var nib = new List<byte>();

        // BOM
        nib.Add(EscTier1); nib.Add(0x0); nib.Add(0x2);

        for (int i = 0; i < s.Length; )
        {
            int cp = char.ConvertToUtf32(s, i);
            i += char.IsHighSurrogate(s[i]) ? 2 : 1;
            EncodeCP((uint)cp, nib);
        }

        if (nib.Count % 2 != 0) nib.Add(0x0);

        // Pack nibbles → bytes
        var bytes = new byte[nib.Count / 2];
        for (int i = 0; i < nib.Count; i += 2)
            bytes[i / 2] = (byte)(((nib[i] & 0xF) << 4) | (nib[i + 1] & 0xF));

        return bytes;
    }

    // ─── Decoder ──────────────────────────────────────────────────────────────

    public static string Decode(byte[] buf)
    {
        // Strip legacy magic if present
        int byteOffset = 0;
        if (buf.Length >= LegacyMagic.Length)
        {
            bool hasMagic = true;
            for (int k = 0; k < LegacyMagic.Length && hasMagic; k++)
                if (buf[k] != LegacyMagic[k]) hasMagic = false;
            if (hasMagic) byteOffset = LegacyMagic.Length;
        }

        // Unpack bytes → nibbles
        var nib = new List<byte>();
        for (int k = byteOffset; k < buf.Length; k++)
        {
            nib.Add((byte)((buf[k] >> 4) & 0xF));
            nib.Add((byte)(buf[k] & 0xF));
        }

        var sb      = new StringBuilder();
        int i       = 0;
        bool capNext = false;
        int len     = nib.Count;

        // Skip BOM
        if (len >= 3 && nib[0] == EscTier1 && nib[1] == 0x0 && nib[2] == 0x2) i = 3;

        void AppendCP(uint cp, ref bool cap)
        {
            string s = char.ConvertFromUtf32((int)cp);
            if (cap) { s = s.ToUpperInvariant(); cap = false; }
            sb.Append(s);
        }

        while (i < len)
        {
            if (i == len - 1 && nib[i] == 0x0) break;
            byte n = nib[i];

            if (n <= 0xB)
            {
                string c = Tier0Chars[n].ToString();
                if (capNext) { c = c.ToUpperInvariant(); capNext = false; }
                sb.Append(c);
                i++; continue;
            }

            if (n == EscCap)
            {
                byte ctrl = (byte)(nib[i + 1] & 0x3);
                if (ctrl == CtrlCap) capNext = true;
                i += 2; continue;
            }

            if (n == EscTier1)
            {
                if (i + 2 >= len) { sb.Append('\uFFFD'); i++; continue; }
                uint cp = ((uint)nib[i+1] << 4) | nib[i+2];
                AppendCP(cp, ref capNext);
                i += 3; continue;
            }

            if (n == EscTier23)
            {
                if (i + 1 >= len) { sb.Append('\uFFFD'); i++; continue; }
                if (nib[i+1] <= 0x7)
                {
                    if (i + 4 >= len) { sb.Append('\uFFFD'); i++; continue; }
                    uint cp = ((uint)nib[i+1] << 12) | ((uint)nib[i+2] << 8)
                            | ((uint)nib[i+3] << 4)  |  nib[i+4];
                    AppendCP(cp, ref capNext);
                    i += 5;
                }
                else
                {
                    if (i + 8 >= len) { sb.Append('\uFFFD'); i++; continue; }
                    uint cp = ((uint)nib[i+2] << 24) | ((uint)nib[i+3] << 20)
                            | ((uint)nib[i+4] << 16) | ((uint)nib[i+5] << 12)
                            | ((uint)nib[i+6] << 8)  | ((uint)nib[i+7] << 4)
                            |  nib[i+8];
                    AppendCP(cp, ref capNext);
                    i += 9;
                }
                continue;
            }

            if (n == EscTier4)
            {
                if (i + 15 >= len) { sb.Append('\uFFFD'); i++; continue; }
                ulong cp = 0;
                for (int j = 1; j <= 15; j++) cp = (cp << 4) | nib[i + j];
                AppendCP((uint)cp, ref capNext);
                i += 16; continue;
            }

            sb.Append('\uFFFD');
            i++;
        }

        return sb.ToString();
    }

    // ─── Utilities ────────────────────────────────────────────────────────────

    public static string EncodeToHex(string s)
    {
        var bytes = Encode(s);
        return string.Join(" ", Array.ConvertAll(bytes, b => b.ToString("x2")));
    }

    public static int PayloadByteSize(string s)
    {
        var nib = new List<byte>();
        for (int i = 0; i < s.Length; )
        {
            int cp = char.ConvertToUtf32(s, i);
            i += char.IsHighSurrogate(s[i]) ? 2 : 1;
            EncodeCP((uint)cp, nib);
        }
        if (nib.Count % 2 != 0) nib.Add(0);
        return nib.Count / 2;
    }

    // ─── Test Suite ───────────────────────────────────────────────────────────

    public static void RunTests()
    {
        var tests = new (string input, string desc)[]
        {
            ("hello",             "lowercase hello"),
            ("Hello",             "capitalized Hello"),
            ("",                  "empty string"),
            ("the rain in spain", "tier0-heavy sentence"),
            ("Hello, World!",     "mixed ascii"),
            ("Héllo",             "accented char (Tier2)"),
            ("日本語",             "Japanese (Tier2)"),
        };

        Console.OutputEncoding = Encoding.UTF8;
        Console.WriteLine("UTF-64 C# Test Suite");
        Console.WriteLine(new string('=', 40));

        foreach (var (input, desc) in tests)
        {
            byte[] enc = Encode(input);
            string dec = Decode(enc);
            bool ok    = dec == input;
            string hex = string.Join(" ", Array.ConvertAll(enc, b => b.ToString("x2")));

            Console.WriteLine($"[{(ok ? "PASS" : "FAIL")}] {desc}");
            if (!ok)
            {
                Console.WriteLine($"  expected: {input}");
                Console.WriteLine($"  got:      {dec}");
            }
            else
            {
                Console.WriteLine($"  \"{input}\" → {PayloadByteSize(input)} bytes");
            }
            Console.WriteLine($"  hex: {hex}\n");
        }
    }
}

class Utf64Runner
{
    static void Main() => Utf64.RunTests();
}

File: utf64.rb
# frozen_string_literal: true
#
# UTF-64 — Universal Transformation Format 64
# Specification v0.2 — Ruby Reference Implementation
#
# Changes from v0.1:
#   - Magic bytes removed from encoder output.
#     Decoder still accepts (and strips) legacy magic-prefixed streams.

# ─── Constants ────────────────────────────────────────────────────────────────

LEGACY_MAGIC = "&-9EX".b.bytes.freeze

TIER0_CHARS = [' ', 'e', 't', 'a', 'o', 'i', 'n', 's', 'h', 'r', 'l', "\n"].freeze
TIER0_MAP   = TIER0_CHARS.each_with_index.to_h.freeze

ESC_CAP    = 0xC
ESC_TIER1  = 0xD
ESC_TIER23 = 0xE
ESC_TIER4  = 0xF
CTRL_CAP   = 0x0

# ─── Encoder ──────────────────────────────────────────────────────────────────

def encode_tier3(cp, nib)
  nib << ESC_TIER23 << 0x8
  24.step(0, -4) { |shift| nib << ((cp >> shift) & 0xF) }
end

def encode_cp(cp, nib)
  char  = cp.chr(Encoding::UTF_8)
  lower = char.downcase
  is_upper = char != lower
  t0l = TIER0_MAP[lower]

  if is_upper && t0l
    nib << ESC_CAP << CTRL_CAP << t0l
    return
  end

  if (t0 = TIER0_MAP[char])
    nib << t0
    return
  end

  if cp <= 0xFF
    nib << ESC_TIER1 << ((cp >> 4) & 0xF) << (cp & 0xF)
    return
  end

  if cp <= 0xFFFF
    n3 = (cp >> 12) & 0xF
    if n3 <= 0x7
      nib << ESC_TIER23 << n3 << ((cp >> 8) & 0xF) << ((cp >> 4) & 0xF) << (cp & 0xF)
    else
      encode_tier3(cp, nib)
    end
    return
  end

  if cp <= 0xFFFFFFFF
    encode_tier3(cp, nib)
    return
  end

  # Tier 4
  nib << ESC_TIER4
  56.step(0, -4) { |shift| nib << ((cp >> shift) & 0xF) }
end

def utf64_encode(s)
  nib = []

  # BOM
  nib << ESC_TIER1 << 0x0 << 0x2

  s.each_codepoint { |cp| encode_cp(cp, nib) }

  nib << 0x0 if nib.length.odd?

  # Pack nibbles → bytes
  nib.each_slice(2).map { |hi, lo| ((hi & 0xF) << 4) | (lo & 0xF) }.pack('C*')
end

# ─── Decoder ──────────────────────────────────────────────────────────────────

def utf64_decode(buf)
  bytes = buf.bytes

  # Strip legacy magic if present
  byte_offset = bytes.first(LEGACY_MAGIC.length) == LEGACY_MAGIC ? LEGACY_MAGIC.length : 0

  # Unpack bytes → nibbles
  nib = bytes[byte_offset..].flat_map { |b| [(b >> 4) & 0xF, b & 0xF] }

  result   = ''.encode(Encoding::UTF_8)
  i        = 0
  cap_next = false
  len      = nib.length

  # Skip BOM
  i = 3 if len >= 3 && nib[0] == ESC_TIER1 && nib[1] == 0x0 && nib[2] == 0x2

  while i < len
    break if i == len - 1 && nib[i] == 0x0

    n = nib[i]

    if n <= 0xB
      c = TIER0_CHARS[n].dup
      if cap_next
        c = c.upcase
        cap_next = false
      end
      result << c
      i += 1
      next
    end

    if n == ESC_CAP
      ctrl = nib[i + 1] & 0x3
      cap_next = true if ctrl == CTRL_CAP
      i += 2
      next
    end

    if n == ESC_TIER1
      if i + 2 >= len
        result << "\uFFFD"
        i += 1
        next
      end
      cp = (nib[i+1] << 4) | nib[i+2]
      c  = cp.chr(Encoding::UTF_8)
      if cap_next
        c = c.upcase
        cap_next = false
      end
      result << c
      i += 3
      next
    end

    if n == ESC_TIER23
      if i + 1 >= len
        result << "\uFFFD"
        i += 1
        next
      end
      if nib[i+1] <= 0x7
        if i + 4 >= len
          result << "\uFFFD"
          i += 1
          next
        end
        cp = (nib[i+1] << 12) | (nib[i+2] << 8) | (nib[i+3] << 4) | nib[i+4]
        c  = cp.chr(Encoding::UTF_8)
        if cap_next
          c = c.upcase
          cap_next = false
        end
        result << c
        i += 5
      else
        if i + 8 >= len
          result << "\uFFFD"
          i += 1
          next
        end
        cp = (nib[i+2] << 24) | (nib[i+3] << 20) | (nib[i+4] << 16) |
             (nib[i+5] << 12) | (nib[i+6] << 8)  | (nib[i+7] << 4)  | nib[i+8]
        result << cp.chr(Encoding::UTF_8)
        i += 9
      end
      next
    end

    if n == ESC_TIER4
      if i + 15 >= len
        result << "\uFFFD"
        i += 1
        next
      end
      cp = (1..15).reduce(0) { |acc, j| (acc << 4) | nib[i + j] }
      result << cp.chr(Encoding::UTF_8)
      i += 16
      next
    end

    result << "\uFFFD"
    i += 1
  end

  result
end

# ─── Utilities ────────────────────────────────────────────────────────────────

def utf64_encode_to_hex(s)
  utf64_encode(s).bytes.map { |b| b.to_s(16).rjust(2, '0') }.join(' ')
end

def utf64_payload_byte_size(s)
  nib = []
  s.each_codepoint { |cp| encode_cp(cp, nib) }
  nib << 0 if nib.length.odd?
  nib.length / 2
end

# ─── Test Suite ───────────────────────────────────────────────────────────────

def run_tests
  tests = [
    ['hello',             'lowercase hello'],
    ['Hello',             'capitalized Hello'],
    ['',                  'empty string'],
    ['the rain in spain', 'tier0-heavy sentence'],
    ['Hello, World!',     'mixed ascii'],
    ['Héllo',             'accented char (Tier2)'],
    ['日本語',             'Japanese (Tier2)'],
  ]

  puts "UTF-64 Ruby Test Suite"
  puts '=' * 40
  tests.each do |input, desc|
    enc  = utf64_encode(input)
    dec  = utf64_decode(enc)
    ok   = dec == input
    hex  = enc.bytes.map { |b| b.to_s(16).rjust(2, '0') }.join(' ')
    puts "[#{ok ? 'PASS' : 'FAIL'}] #{desc}"
    if !ok
      puts "  expected: #{input.inspect}"
      puts "  got:      #{dec.inspect}"
    else
      puts "  #{input.inspect} → #{utf64_payload_byte_size(input)} bytes"
    end
    puts "  hex: #{hex}\n\n"
  end
end

run_tests
